### ISSUE TITLE
ContentContainer scoped URL Rules

### ISSUE THREAD

**Issue by luke-** [humhub/humhub#5151]
Currently it is not possible for a ContentContainer (Space / User) based URL Rules.
Ideally these could be defined like normal `urlMangerRules` in `config.php` of a module.

See possible config below:

```php
return [
    'id' => 'space',
    'class' => Module::class,
    //...
    'urlManagerRules' => [
        ['class' => 'humhub\modules\space\components\UrlRule'],
        'spaces' => 'space/spaces',

        // ****** New Rule Ideas Below *********** //

        // for:  /s/example/space/space/home => /s/example/home
        '<contentContainer>/home' => 'space/space/home'

        // for complexer rules (maybe bases on  `humhub\modules\content\components\BaseUrlRule`
        ['class' => 'humhub\modules\wiki\components\UrlRule'],

    ],
];
```
Of course, performance and caching is very important .

**Comment by luke-** [comment]
See also: https://github.com/humhub/wiki/issues/136

**Comment by ArchBlood** [comment]
I completely agree that this should be done!

Currently doing the same thing in the Discussions module
```php
    'urlManagerRules' => [
        'discussions/index/' => 'discussions/default/index',
        'discussions/guest/' => 'discussions/default/guest',
        'discussions/create/' => 'discussions/default/create',
        'discussions/view' => 'discussions/default/view',
        'discussions/update' => 'discussions/default/update',
        'discussions/delete' => 'discussions/default/delete'
    ],
```
Although I'm not working with a `class` to define just using `urlManagerRules` as is, or should I think of doing the same? 🤔

**Comment by yurabakhtin** [comment]
@luke- For info: We have an URL like this `http://humhub.local/s/space-url/wiki/page/view?title=Wiki+Page+title` and we want to create an Alias like `http://humhub.local/s/space-url/wiki/Wiki+Page+title` (or `http://humhub.local/s/space-url/wiki/wiki-page-title` when Wiki Pages will have some new param page url, or if we will convert automatically the page title "Wiki Page title" to page URL `wiki-page-title`).

> `// for complexer rules (maybe bases on  humhub\modules\content\components\BaseUrlRule`
> `['class' => 'humhub\modules\wiki\components\UrlRule'],`

If we will add a new rule in the config `urlManagerRules` at the end, then it will be applied/called only if all previous/top rules don't detect own rule.
For example, if we call URL `/s/space-url/wiki/...` then it is handled here https://github.com/humhub/humhub/blob/master/protected/humhub/modules/space/components/UrlRule.php#L60-L79, so if space is found by the `space-url` we don't allow to parse url by next rules(because `return false;`), i.e. the URL `/s/space-url/wiki/...` will not be parsed by next rule `['class' => 'humhub\modules\wiki\components\UrlRule']`, this behaviour is in yii core class `yii\web\UrlManager`.
But if we will call URL like `/wiki/...` then it may be parsed by the next rule `humhub\modules\wiki\components\UrlRule` because the space rule parse only URL started with `/s/...`. And as you can see this way works on modules like example of `discussions` above, or example with "class" - https://github.com/humhub/custom-pages/blob/master/config.php#L19.

---

Currently when we open URL `http://humhub.local/s/space-url/wiki/Wiki+Page+title` we get "Page not found.", I think it is displayed by `protected/humhub/controllers/ErrorController.php`. What if all not found pages from modules will be processed by own ErrorController e.g. `protected/modules/wiki/controllers/ErrorController.php` and then inside we will try to find a Wiki Page by title/url?

**Comment by luke-** [comment]
Thank you for the info!

@ArchBlood  The Discussion module should not be affected by that, because it does not work as Space module or?

Can we solve the problem from the approach e.g. like this?

- Migrate existing `SpaceUrlRule` and `UserUrlRule` into one `ContentContainerUrlRule` to ease handling

- The ModuleAutoloader which is parsing the `UrlMangerRules` in den `config.php` should **move** all rules which contains a `<contentContainer>` keyword or classes extended by `CustomContentContainerUrlRule`  to  a `ContentContainerUrlRule` static attribut.

- The current Yii2 UrlManager/Route Parser only run/checks the `ContentContainerUrlRule` once and rules without any ContentContainer flag/class. (e.g. /dashboard)

- When `ContentContainerUrl` found a valid container (`/s/example` or `/u/john`) it first tries to resolve additional rules from attached previously by ModuleLoader before using the `defaultRoute`.  
  
@yurabakhtin What do you think?

**Comment by ArchBlood** [comment]
We're currently being asked to work in a Space function, we're still processing if we want to do this though

**Comment by yurabakhtin** [comment]
@luke- 

> What do you think?

> - Migrate existing `SpaceUrlRule` and `UserUrlRule` into one `ContentContainerUrlRule` to ease handling

Maybe no need to migrate because they have many different code.

> - The ModuleAutoloader which is parsing the `UrlMangerRules` in den `config.php` should move all rules which contains a `<contentContainer>` keyword or classes extended by `CustomContentContainerUrlRule` to a `ContentContainerUrlRule` static attribut.

I didn't investigate deeply how it is possible, but ok if you think it can be implemented. Maybe it is possible to handle rules `<contentContainer>` from the `SpaceUrlRule` and `UserUrlRule`.

**Comment by luke-** [comment]
@yurabakhtin Ok, do you just like to check out what the options are here? 

We probably need to store the ContentContainerUrlRules from the module configs in some special place where UserUrlRule and SpaceUrlRule can access them.  

Maybe both classes could be extended by AbstractContainerUrlRule which holds such UrlRules.

**Comment by yurabakhtin** [comment]
@luke- Please review how I implemented a solution in core PR https://github.com/humhub/humhub/pull/5152 and how it is used from module side in PR https://github.com/humhub/wiki/pull/189.

**Comment by yurabakhtin** [comment]
@luke- After first changes I fixed the broken tests https://github.com/humhub/wiki/actions/runs/1001105375 in commits https://github.com/humhub/humhub/pull/5152/commits/46bc773f521027def31accbfc961a850f6f34f49 and https://github.com/humhub/wiki/pull/189/commits/fe8938f72a59044c4440d1a8bb22afbdd7101706, but wiki module still has the broken tests but only for API tests - https://github.com/humhub/wiki/runs/2990613839?check_suite_focus=true#step:25:34, however for me it works:

<img width="610" alt="wiki_api_tests" src="https://user-images.githubusercontent.com/6995524/124481519-0a7f5900-ddb1-11eb-9e47-c7af28ec29fe.png">

Maybe the REST API module is not loaded properly on running the tests but I don't see any problems there https://github.com/humhub/wiki/runs/2990613839?check_suite_focus=true#step:18:18.

**Comment by luke-** [comment]
@yurabakhtin I like the solution with the interface. Do you think we can somehow allow simpler rules without a custom class?

**Comment by yurabakhtin** [comment]
@luke- 

> Do you think we can somehow allow simpler rules without a custom class?

Ok, I will investigate this, but please note I implemented that with class way in order to check the Wiki Page exists in DB by the requested title.

**Comment by luke-** [comment]
@yurabakhtin The solution with the wiki page and the possibility to create a custom class is great! 

For some cases a solution without extra class would be useful.

E.g.
https://community.humhub.com/s/feature-requests/space/space/about should be
https://community.humhub.com/s/feature-requests/about

would be great if we can find a lightweight approach for such cases.

**Comment by yurabakhtin** [comment]
@luke- Commit https://github.com/humhub/humhub/pull/5152/commits/5f157243e88e837cf97d0693f112e44334a1a231:

Please find new rules in config files of the space and user modules:

use config:
- `'<userContainer>/home' => 'user/profile/home'`
- `'<userContainer>/about' => 'user/profile/about'`

space config:
- `'<spaceContainer>/home' => 'space/space/home'`
- `'<spaceContainer>/about' => 'space/space/about'`

As you can see I implemented new `<userContainer>` and `<spaceContainer>`, but also it may works with `<contentContainer>` however there is a conflict if we will use same rule key from different module configs like this example:

user config:
- `'<contentContainer>/home' => 'user/profile/home'`

space config:
- `'<contentContainer>/home' => 'space/space/home'`

In such case the user routing will be tried to processed by space routing and we will get an error. It is why I decided for such case to implement a different rule prefixes `<userContainer>` and `<spaceContainer>`.

---

We can use a rule like this `'<contentContainer>/wiki/<title>' => 'wiki/page/view',` from external module config and it will works to generate a proper URLs and open the wiki page by requested title, but I don't recommend this rule because in such case we cannot check the wiki page by requested page really exists in DB, so if we will open an URL like `/s/welcome-space/wiki/page` then we cannot go to expected index page of all wiki pages, because this URL will be processed by rule `<contentContainer>/wiki/<title>` and will go to wiki view page with title == 'page'. In order to avoid such conflict we should use currently implemented way with class rule `['class' => 'humhub\modules\wiki\components\WikiPageUrlRule']`, but such way has too a small issue when we have a wiki page with title == 'page'.
Should we find a solution for this? Maybe it can be fixed by rule like `'<contentContainer>/wiki/page:<title>' => ...` or `'<contentContainer>/wiki-page/<title>' => ...`, i.e. try to set URL without existing module/controller/action names.

**Comment by lisandi** [comment]
why not simply using a simple "w" instead of wiki - Making URLs shorter is much more useful and much easier to navigate and to remember.
u - user
p - page
w - wiki
g - group
d - discussion
s - space
...

also the doubling of space/space/.. or wiki/page/.. could be shortened for usability.

**Comment by luke-** [comment]
@lisandi Yes, this issue is about shortening and improving URLs within a space/user. See more here: https://github.com/humhub/wiki/issues/136#issuecomment-882147326
